### PR TITLE
feat(actions): git-aware cut / paste / rename fs actions

### LINKED ISSUE #2474: Git Actions
https://github.com/nvim-tree/nvim-tree.lua/blob/aaee4cd896b74f85a81bed7eef2db7869960c4d0/lua/nvim-tree/actions/fs/rename-file.lua#L23-L40

### **Describe the solution you'd like**
The  `e` and `r` don't respect git repo: by renaming something we loose track of file path history. I would like to be able to rename filepath by using `git mv` (instead of system command) in order to keep track of file renames:
```
git mv <src_path> <dst_path>
```

Currently, `nvim-tree` doesn't provide this feature forcing me to rename file by using git manually.

### **Describe alternatives you've considered**
Running `!git mv % <path_no_new_name>` is rather clunky...
Running `!git -A <path_no_new_name>` is rather clunky and not something I want...

### **Can this functionality be implemented utilising API?**
NO.

### **Additional context**
I'ts difficult to easily contribute API feature, cause:
1. The [runner](https://github.com/nvim-tree/nvim-tree.lua/blob/aaee4cd896b74f85a81bed7eef2db7869960c4d0/lua/nvim-tree/git/runner.lua#L169-L173) is not a generic lua wrapper for git:
* it's hard-wired to run only single git subcommand: `status`
* quite intricate interally - it's hard to extend
* no docs/comments

3. This repo has no dedicated `git` wrapper for running & handling git output (there is a dedicated fast [libgit2 lua bindings](https://github.com/libgit2/luagit2) btw)
4. Intent of `nvim-tree.git` is unclear

### PR DESCRIPTION
**Pull request by hinell**
https://github.com/nvim-tree/nvim-tree.lua/assets/8136158/a46aa792-b6d6-41fb-9af6-679e918ef5c9

> git-aware file renames right in your nvim-tree; refactor your source tree easily
> NOTE: Demo uses rebound default `x,c,p,r,e,u,<c-r>` keymap bindings
> NOTE: Letter `R` signifies git renamed file (setup by `confi.renderer.icons.glyphs.git`

This PR allows:
* Moving/renaming git-tracked files by using `fs.*` like file actions for cut / paste / rename
*  Moving/renaming non-tracked files by `fs.{cut,paste}` as usual
* Wraps around `fs.*` APIs nicely, minimize redundancy
* Utilizes log / info utils
* Closes #2474
### Usage

No keymaps setup by default. Example:
```lua
-- $HOME/.config/nvim/lua/user/nvim-keymaps.lua
-- put this inside autocmd callback for NvimTree filetype, see `h autocmd` and `h nvim_create_autocmd()`
local opts = { buffer = autoCmdEvent.buf, nowait = true }
local api  = require "nvim-tree.api"
vim.keymap.set("n", "x"    , api.git.cut            , vim.tbl_extend("keep", { desc = "nvim-tree: git: cut file into clipboard"   }, opts)
vim.keymap.set("n", "p"    , api.git.paste          , vim.tbl_extend("keep", { desc = "nvim-tree: git: past file from clipboard"  }, opts)
vim.keymap.set("n", "r"    , api.git.rename         , vim.tbl_extend("keep", { desc = "nvim-tree: git: rename: filename"          }, opts)
vim.keymap.set("n", "e"    , api.git.rename_basename, vim.tbl_extend("keep", { desc = "nvim-tree: git: rename: filename base"     }, opts)
vim.keymap.set("n", "u"    , api.git.rename_absolute, vim.tbl_extend("keep", { desc = "nvim-tree: git: rename: filename absolute" }, opts)
vim.keymap.set("n", "<C-r>", api.git.rename_sub     , vim.tbl_extend("keep", { desc = "nvim-tree: git: rename: omit filename"	  }, opts)
```

### API
* Adds:    
`api.git.cut(node)` - store a file under cursor into a git actions clipboard (just like `api.fs.cut`)
`api.git.paste(node)` - pop files from clipboard into a folder under cursor
`api.git.rename(node)` - rename file under cursor
`api.git.rename_sub(node)`
`api.git.rename_basename(node)`
`api.git.rename_absolute(node)`
`api.git.clear_clipboard(node)`
`api.git.print_clipboard(node)`

* Utils: 
`git(uvSpawnOptions, onErrOrData, onExit)` - a wrapper func around `git` cli command (uses `uv`)
`git_is_tracked(path, cb, onExit)` - checks whether given path is tracked  
`git_mv(path_src, path_dst, callback, onExit)` -  moves files around by using git 
`Clipboard` - a class for future for clipboard; used for git actions local copy / paste stack

### Design considerations

I'm aware of [git runner for git](https://github.com/nvim-tree/nvim-tree.lua/blob/46e1f776f0a714099b9160604fbb62fba2540376/lua/nvim-tree/git/runner.lua#L169-L183) but as I said earlier elsewhere - it's too limited to be used for a general purpose git commands, so this PR introduces a full-fledged proxy to git.

### CODE REVIEW

File: lua/nvim-tree/actions/git/cli.lua
Comment (alex-courtis) [inline lua/nvim-tree/actions/git/cli.lua:4]: Oh wow... thanks for the alert!

File: doc/nvim-tree-lua.txt
Comment (alex-courtis) [inline doc/nvim-tree-lua.txt:1919]: We don't need this paragraph to justify; the reasons and benefits are clear.

File: doc/nvim-tree-lua.txt
Comment (alex-courtis) [inline doc/nvim-tree-lua.txt:1911]: Great work.

File: lua/nvim-tree/actions/git/init.lua
Comment (alex-courtis) [inline lua/nvim-tree/actions/git/init.lua:43]: /nit please use snake case rather than camel case.

File: lua/nvim-tree/actions/git/cli.lua
Comment (alex-courtis) [inline lua/nvim-tree/actions/git/cli.lua:34]: This functionality appears to duplicate `git/runner.lua`.

Let's not repeat ourselves and share that code.

File: lua/nvim-tree/actions/git/init.lua
Comment (alex-courtis) [inline lua/nvim-tree/actions/git/init.lua:197]: Very nice, thank you.

File: lua/nvim-tree/actions/git/init.lua
Comment (alex-courtis) [inline lua/nvim-tree/actions/git/init.lua:154]: goto? no thank you. Standard flow of control please.

File: lua/nvim-tree/actions/git/clipboard.lua
Comment (alex-courtis) [inline lua/nvim-tree/actions/git/clipboard.lua:7]: The prototype pattern is surfacing a lot of warnings and confusing lua_language_server. Any reason we need that over an "ordinary" class?

File: lua/nvim-tree/actions/git/init.lua
Comment (alex-courtis) [inline lua/nvim-tree/actions/git/init.lua:232]: This method repeats a lot of `rename-file.lua`.

Can we please share?

File: lua/nvim-tree/actions/git/init.lua
Comment (alex-courtis) [inline lua/nvim-tree/actions/git/init.lua:318]: What is calling this?

**Review by alex-courtis — CHANGES_REQUESTED** [review]
This is a fantastic! I love the direction you are taking with the fallback pattern, which won't disturb current users.

I think a lot of users, myself included, will end up mapping to these actions instead of the defaults.

This one will take a while, I'd like to dogfood this for some time before merge.

Big concern: repetition -> maintainability. There's a lot of duplicated code here. Suggestions:

1. Fold `actions/git/init.lua` `actions/git/init.lua` into the existing `copy-paste.lua`, `rename-file.lua` etc, extending the current persistence mechanisms as needed.
2. Use the existing `git/runner.lua`, exposing internal API / refactoring as necessary.
3. There's a lot of great improvements here that should be applied to existing code. Not in this PR. One of:
    - Prefactor PR for existing `copy-paste.lua` etc. before this PR
    - Tidy / improvement PR after merging this PR

**Comment by alex-courtis** [comment]
ALTERNATIVE:

This isn't particularly tightly coupled with nvim-tree. It could easily be extracted into an [extension plugin](https://github.com/nvim-tree/nvim-tree.lua/wiki/Extension-Plugins).

What do you reckon @gegoune @Akmadan23 ?

**Comment by Akmadan23** [comment]
> This isn't particularly tightly coupled with nvim-tree.

That's true, but not completely... Nvim-tree already has some level of git integration, so it makes sense to enhance this aspect, and having these features ready out of the box would be really handy.

**Comment by hinell** [comment]
> _It could easily be extracted into an [extension plugin](https://github.com/nvim-tree/nvim-tree.lua/wiki/Extension-Plugins)._

@alex-courtis  I agree with @Akmadan23 - git support out of the box is essential.

I'm going to do some refactoring. Many modules in this project simply don't export routines that can be reused so I was forced to duplicate it. I'm under impression that this it wasn't intended to be modular in first place or author's werent even suspecting that it might be reusable.

As you can see in this PR I've exported everything to make `nvim-tree` git actions highly extensible. We can even use them in some other plugins like `lualine` e.g. to check clipboard status.

I also plan to introduce simple undo history in the future, when I have enough time on my hands.

**Comment by hinell** [comment]
> _It could easily be extracted into an [extension plugin](https://github.com/nvim-tree/nvim-tree.lua/wiki/Extension-Plugins)._

@alex-courtis  I agree with @Akmadan23 - git support out of the box is essential.

I'm going to do some refactoring. Many modules in this project simply don't export routines that can be reused so I was forced to duplicate it. I'm under impression that the code was NOT intended to be modular in first place or authors werent even suspecting that the code might be used elsewhere.

As you can see in this PR I've exported everything to make `nvim-tree` git actions highly extensible. We can even use parts of it in some other plugins like `lualine` e.g. to check clipboard status.

I also plan to introduce simple undo history in the future, when I have enough time on my hands.

File: lua/nvim-tree/actions/git/init.lua
Comment (hinell) [reply lua/nvim-tree/actions/git/init.lua:232]: It can be shared by a separate `refactor(actions): rename-file.lua` PR. Are you going to accept a separate PR with `rename-file.lua` refactoring? 

Currently `rename-file.lua` isn't exporting internal utils so it's not reusable. The chunk of code above si similar to (IMO very inefficient) following string comparison:
https://github.com/nvim-tree/nvim-tree.lua/blob/46e1f776f0a714099b9160604fbb62fba2540376/lua/nvim-tree/actions/fs/rename-file.lua#L64-L79

**Comment by hinell** [comment]
> _It could easily be extracted into an [extension plugin](https://github.com/nvim-tree/nvim-tree.lua/wiki/Extension-Plugins)._

@alex-courtis  I agree with @Akmadan23 - git support out of the box is essential.

I'm going to do some refactoring. Many modules in this project simply don't export routines that can be reused so I was forced to duplicate the. I'm under impression that the code was NOT intended to be modular in first place or authors werent even suspecting that the code might be used elsewhere.

As you can see in this PR I've exported everything to make `nvim-tree` git actions highly extensible. We can even use parts of it in some other plugins like `lualine` e.g. to check clipboard status.

I also plan to introduce simple undo history in the future, when I have enough time on my hands.

**Comment by hinell** [comment]
> _Fold actions/git/init.lua actions/git/init.lua into the existing copy-paste.lua, rename-file.lua etc, extending the current persistence mechanisms as needed._

By nature, both `fs` and `git` actions are different. It's waaay better to keep them separate. In future, git actions may be extended to more git commands or even see integration with others plugins.

I would suggest to to put common / reusable code into exported functions or classes. This will result in less code and easier maintance.

**Comment by hinell** [comment]
> _It could easily be extracted into an [extension plugin](https://github.com/nvim-tree/nvim-tree.lua/wiki/Extension-Plugins)._

@alex-courtis  I agree with @Akmadan23 - support for git commands out of the box is essential.

I'm going to do some refactoring. Many modules in this project simply don't export routines that can be reused so I was forced to duplicate them. I'm under impression that the code was NOT intended to be modular/reusable in first place or authors werent even suspecting that the code might be used elsewhere.

As you can see in this PR I've exported everything to make `nvim-tree` git actions highly reusable. We can even use parts of it in some other plugins like `lualine` e.g. to check clipboard status.

I also plan to introduce simple undo history in the future, when I have enough time on my hands.

### Mergin with `fs` actions

> _Fold actions/git/init.lua actions/git/init.lua into the existing copy-paste.lua, rename-file.lua etc, extending the current persistence mechanisms as needed._

By nature, both `fs` and `git` actions are different. It's waaay better to keep them separate. In future, git actions may be extended to more git commands or even see integration with others plugins.

My suggestion is to put and exported common / reusable code into functions or classes. This will decrease amount of code needed to maintain.

**Comment by hinell** [comment]
> _It could easily be extracted into an [extension plugin](https://github.com/nvim-tree/nvim-tree.lua/wiki/Extension-Plugins)._

@alex-courtis  I agree with @Akmadan23 - support for git commands out of the box is essential.

I'm going to do some refactoring. Many modules in this project simply don't export routines that can be reused so I was forced to duplicate them. I'm under impression that the code was NOT intended to be modular/reusable in first place or authors werent even suspecting that the code might be used elsewhere.

As you can see in this PR I've exported everything to make `nvim-tree` git actions highly reusable. We can even use parts of it in some other plugins like `lualine` e.g. to check clipboard status.

I also plan to introduce simple undo history in the future, when I have enough time on my hands.

### Mergin with `fs` actions

> _Fold actions/git/init.lua actions/git/init.lua into the existing copy-paste.lua, rename-file.lua etc, extending the current persistence mechanisms as needed._

By nature, both `fs` and `git` actions are different. It's waaay better to keep them separate. In future, git actions may be extended to more git commands or even see integration with others plugins.

My suggestion is to put and exported common / reusable code into functions or classes. This will decrease amount of code needed to maintain.

**Comment by hinell** [comment]
> _It could easily be extracted into an [extension plugin](https://github.com/nvim-tree/nvim-tree.lua/wiki/Extension-Plugins)._

@alex-courtis  I agree with @Akmadan23 - support for git commands out of the box is essential.

I'm going to do some refactoring. Many modules in this project simply don't export routines that can be reused so I was forced to duplicate them. I'm under impression that the code was NOT intended to be modular/reusable in first place or authors werent even suspecting that the code might be used elsewhere.

As you can see in this PR I've exported everything to make `nvim-tree` git actions highly reusable. We can even use parts of it in some other plugins like `lualine` e.g. to check clipboard status.

I also plan to introduce simple undo history in the future, when I have enough time on my hands.

### Merging with `fs` actions

> _Fold actions/git/init.lua actions/git/init.lua into the existing copy-paste.lua, rename-file.lua etc, extending the current persistence mechanisms as needed._

By nature, both `fs` and `git` actions are different. It's waaay better to keep them separate. In future, git actions may be extended to more git commands or even see integration with others plugins.

My suggestion is to put and exported common / reusable code into functions or classes. This will decrease amount of code needed to maintain.

**Comment by hinell** [comment]
## On why this is not a plugin

tl;dr: `nvim-tree` doesn't provide any sane API for plugins. 

It would be nice if `nvim-tree` borrows some `nvim-neo-tree` idea of Sources™, but in a sense of plugins. You can take a look how plugins for `nvim-treesitter` or `telescope.nvim` or `nvim-cmp` are managed. `nvim-tree` sadly is nowhere close.

If `nvim-tree` provided a generic UI Tree Widget along with standard plugins like `lsp`, `fs`, which would essentially provide various tweaks over default UI Tree widget, it would greatly simplify architecture.

Like, what if plugins could look like this?: 

```lua
plugin.source     -- a table of funcs that export recursive table to be displayed as a tree;  user can switch sources
plugin.source_ui  -- a table of functions handling UI events emitted by nvim-tree to modify appearance
plugin.commands   -- i.e. a table actions exposed to user / others plugins (e.g. actions over focused node)
plugin.clipboard  -- integration with nvim-tree global clipboard
plugin.history    -- integration with nvim-tree actions history

--- etc 

```

That would make extending nvim-tree very simple. 


Btw, `nvim-neo-tree` has implemented [git `actions`](https://github.com/nvim-neo-tree/neo-tree.nvim/blob/230ff118613fa07138ba579b89d13ec2201530b9/README.md?plain=1#L385-L404) long ago.

File: lua/nvim-tree/actions/git/init.lua
Comment (alex-courtis) [reply lua/nvim-tree/actions/git/init.lua:232]: > It can be shared by a separate `refactor(actions): rename-file.lua` PR. Are you going to accept a separate PR with `rename-file.lua` refactoring?

Absolutely! That's what I meant by "prefactor". Sorry, I shoud have been clearer.

Something like "chore(#2542): refactor rename file"

**Comment by alex-courtis** [comment]
> I also plan to introduce simple undo history in the future, when I have enough time on my hands.

That would indeed be useful; I've reflexively tried that a few times and was disappointed.

**Comment by alex-courtis** [comment]
> That's true, but not completely... Nvim-tree already has some level of git integration, so it makes sense to enhance this aspect, and having these features ready out of the box would be really handy.

> tl;dr: nvim-tree doesn't provide any sane API for plugins.

> As you can see in this PR I've exported everything to make nvim-tree git actions highly reusable. We can even use parts of it in some other plugins like lualine e.g. to check clipboard status.

Fair enough, this functionality is suited for nvim-tree itself rather than another plugin.

RE API: https://github.com/nvim-tree/nvim-tree.lua#roadmap we have been adding API as needed, however that's not practical for this case.

**Comment by alex-courtis** [comment]
> I'm under impression that the code was NOT intended to be modular/reusable in first place or authors werent even suspecting that the code might be used elsewhere.

We are always striving to improve. Modularising / exposing is desirable.

> My suggestion is to put and exporte common / reusable code into functions or classes. This will decrease amount of code needed to be maintained.

Yes please!

File: lua/nvim-tree/actions/git/init.lua
Comment (gegoune) [reply lua/nvim-tree/actions/git/init.lua:232]: Or `refactor(…): …`. :)

Comment (alex-courtis) [reply lua/nvim-tree/actions/git/init.lua:232]: Heh... didn't realise that was a prefix, thank you.

Comment (hinell) [reply lua/nvim-tree/actions/git/init.lua:232]: Please merge:
https://github.com/nvim-tree/nvim-tree.lua/pull/2550

This PR is going to use common code from that PR.

File: lua/nvim-tree/actions/git/cli.lua
Comment (hinell) [reply lua/nvim-tree/actions/git/cli.lua:34]: Please merge:
https://github.com/nvim-tree/nvim-tree.lua/pull/2551

**Comment by hinell** [comment]
Before I can proceed, we need the following merged:

- #2550
- #2551

**Comment by hinell** [comment]
Before I can proceed, I request the following to be merged:

- #2550
- #2551

**Comment by hinell** [comment]
I've refactored this PR to address requested changes. To actually 
lua/nvim-tree/actions/git/init.lua

**Comment by hinell** [comment]
Before I can proceed, I request the following to be merged:

- #2550
- #2551


### UPD: I've refactored this PR to address requested changes.

To actually quickly test this PR, you have to pull down the above PRs and apply changes as patches (no changes in your workint tree are assumed):
```bash
 gh pr checkout 2550
 gh pr checkout 2551

git co actions-git
git checkout -p git-integration-refactor lua/nvim-tree/git/runner.lua
git checkout -p actions-fs-rename lua/nvim-tree/actions/fs/rename-file.lua lua/nvim-tree/utils-ui.lua
git checkout -p actions-fs-rename lua/nvim-tree/api.lua
```

Use `reset --hard` to reset all files.

**Comment by alex-courtis** [comment]
> Before I can proceed, I request the following to be merged:
> 
>     * [refactor(actions): common code for rename-file #2550](https://github.com/nvim-tree/nvim-tree.lua/pull/2550)
> 
>     * [refactor(git): common code to a generic git cll #2551](https://github.com/nvim-tree/nvim-tree.lua/pull/2551)
> 
> 
> ### UPD: I've refactored this PR to address requested changes.
> 
> To actually quickly test this PR, you have to pull down the above PRs and apply changes as patches (no changes in your workint tree are assumed):
> 
> ```shell
>  gh pr checkout 2550
>  gh pr checkout 2551
> 
> git co actions-git
> git checkout -p git-integration-refactor lua/nvim-tree/git/runner.lua
> git checkout -p actions-fs-rename lua/nvim-tree/actions/fs/rename-file.lua lua/nvim-tree/utils-ui.lua
> git checkout -p actions-fs-rename lua/nvim-tree/api.lua
> ```
> 
> Use `reset --hard` to reset all files.

Thanks for the detailed instructions.

Unfortunately my time is limited. I'll process PRs sequentially: #2551 then #2550 then this PR.

**Comment by hinell** [comment]
> Unfortunately my time is limited

That's fine, I'm waiting those PRs to land.

**Comment by gegoune** [comment]
I have unresolved reviews from Alex that @hinell resolved. @hinell once again, review authors are to resolve their comments.

**Comment by alex-courtis** [comment]
See https://github.com/nvim-tree/nvim-tree.lua/pull/2551#issuecomment-1828908354

**Pull request closed without merging**
